### PR TITLE
[FIX] Use LEFT JOIN to fix incorrect search results when using auto_join

### LINKED ISSUE #25175: auto_join incorrect results for ['|', (field.name, '=', 'something'), ('field, '=', False)]
Impacted versions:
8.0 up to Master

Steps to reproduce:
Set auto_join = true for some Many2one field. Combine a subsearch with a search on False.

Current behavior:
Records without the relational field set are not returned

Expected behavior:
Both records without the relational field and records matching the subsearch are returned

Below is an example in which res.partner's `state_id` is set to auto_join. We can search for partners without state_id and we can search partners in Florida but the combined query only returns the partners in Florida.

```
>>> session.env['res.partner'].search([('state_id', '=', False)])
2018-06-10 14:54:17,182 17553 DEBUG 11master odoo.sql_db: query: SELECT "res_partner".id FROM "res_partner" WHERE (("res_partner"."active" = true)  AND  "res_partner"."state_id" IS NULL ) ORDER BY "res_partner"."display_name"  
res.partner(10, 36, 31, 19, 18, 9, 40, 24, 17, 41, 42, 16, 14, 28, 34, 27, 11, 20, 22, 33, 23, 15, 35, 8, 7)

>>> session.env['res.partner'].search([('state_id.name', '=', 'Florida')])
2018-06-10 14:54:50,990 17553 DEBUG 11master odoo.sql_db: query: SELECT "res_partner".id FROM "res_country_state" as "res_partner__state_id","res_partner" WHERE ("res_partner"."state_id"="res_partner__state_id"."id") AND (("res_partner"."active" = true)  AND  ("res_partner__state_id"."name" = 'Florida')) ORDER BY "res_partner"."display_name"  
res.partner(13, 29, 30)

>>> session.env['res.partner'].search(['|', ('state_id', '=', False), ('state_id.name', '=', 'Florida')])
2018-06-10 14:54:06,991 17553 DEBUG 11master odoo.sql_db: query: SELECT "res_partner".id FROM "res_country_state" as "res_partner__state_id","res_partner" WHERE ("res_partner"."state_id"="res_partner__state_id"."id") AND (("res_partner"."active" = true)  AND  ("res_partner"."state_id" IS NULL   OR  ("res_partner__state_id"."name" = 'Florida'))) ORDER BY "res_partner"."display_name"  
res.partner(13, 29, 30)
```

### PR DESCRIPTION
**Pull request by StefanRijnhart**
Fixes #25175

While osv.query.Query supports different kinds of joins,
osv.expression.Expression only supports implicit joins. The solution is to
add the join type in Expression and pass the joins to Query in its native
format.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by tivisse** [comment]
I'm waiting this to be reviewed by @rco-odoo or @odony 

I think that they already had a discussion about this subject and I would like to have their input !

Thanks.

**Comment by StefanRijnhart** [comment]
Currently investigating an issue where an implicit join is generated with this code without the corresponding identity clause, thus producing incorrect filtering results.

**Comment by StefanRijnhart** [comment]
The problem above occurred when evaluating record rules. I fixed this by implementing the `+` and `+=` operators for the Query object and refactor any code that manipulated Query internals. This means that ir.rule's domain_get method now returns a Query object.

**Comment by Yenthe666** [comment]
@odony or @rco-odoo would love your feedback here.

**Comment by StefanRijnhart** [comment]
@ovnicraft Done, thanks for noticing!

File: odoo/osv/query.py [hunk 1]
```diff
@@ -160,15 +160,47 @@ def add_joins_for_table(lhs):
...
+                    self.joins[join_table].append(join)
+        return self
+
+    def __add__(self, query):
+        """ Return a new copy of `self` that includes `query` """
+        result = Query(
+            list(self.tables or []),
+            list(self.where_clause or []),
+            list(self.where_clause_params or []),
+            deepcopy(self.joins or {}))
```
Comment (rco-odoo) [inline odoo/osv/query.py]: if keys and values are immutable, `deepcopy` is probably not necessary

File: odoo/osv/expression.py [hunk 2]
```diff
@@ -1297,9 +1319,6 @@ def to_sql(self):
 
         assert len(stack) == 1
         query = stack[0]
-        joins = ' AND '.join(self.joins)
-        if joins:
-            query = '(%s) AND %s' % (joins, query)
 
         params.reverse()
         return query, params
```
Comment (rco-odoo) [inline odoo/osv/expression.py]: maybe the `expression` object could return a `Query` object with everything in it (tables, joins, where clause and params)

File: odoo/models.py [hunk 3]
```diff
@@ -3041,12 +3041,16 @@ def check_access_rule(self, operation):
...
-                query = query + " AND ".join(where_clause)
+            query = self.env['ir.rule'].domain_get(self._name, operation)
+            if query.where_clause:
+                table = '"%s"' % self._table
                 for sub_ids in self._cr.split_for_in_conditions(self.ids):
-                    self._cr.execute(query, [sub_ids] + where_params)
+                    chunk_query = query + Query([table], [table + '.id IN %s'], [sub_ids])
+                    from_clause, where_clause, params = chunk_query.get_sql()
+                    self.env.cr.execute(
+                        "SELECT DISTINCT({table}.id) FROM {from_clause} WHERE {where_clause}".format(
```
Comment (rco-odoo) [inline odoo/models.py]: Avoid `DISTINCT` in queries (potential performance issue);
it is cheaper to make a `set` with results in Python.

File: odoo/osv/query.py [hunk 4]
```diff
@@ -160,15 +160,47 @@ def add_joins_for_table(lhs):
...
+            if table not in self.tables:
+                self.tables.append(table)
+        for join_table in query.joins:
+            self.joins.setdefault(join_table, [])
+            for join in query.joins[join_table]:
+                if join not in self.joins[join_table]:
+                    self.joins[join_table].append(join)
+        return self
+
+    def __add__(self, query):
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/osv/query.py]: And maybe using `query1 & query2` as an alias...

[the remark this answers is not in the document]
Comment (StefanRijnhart) [reply odoo/osv/query.py]: @rco-odoo thanks again for your thorough review during the ODOO Experience!

I have addressed all comments. One thing I am less happy with is dropping the operator that modifies Query in place. The effect is that we have to return the combined Query object everywhere: https://github.com/odoo/odoo/pull/25182/commits/0577b4e27dad8b17eb2b1750183b26a779cbe71f

This is a dangerous change as well, as it is easy to miss an occurrence of this in any third party code resulting in silently dropped access policies. If you agree, I'd gladly implement the `&=` operator after all.

File: odoo/osv/query.py (hunk 1 above)
Comment (StefanRijnhart) [reply odoo/osv/query.py]: I kept this deepcopy because the values are lists and thus mutable.

File: odoo/models.py (hunk 3 above)
Comment (StefanRijnhart) [reply odoo/models.py]: I noticed that deduplication is carried out in the next call so I simply removed the DISTINCT.

**Comment by mreficent** [comment]
conflicts

**Comment by mreficent** [comment]
@StefanRijnhart ?

**Comment by StefanRijnhart** [comment]
There are some open issues with this patch. I don't have time to pursue this now.

**Pull request closed without merging**